packages/wallet/src/wallet.ts

```diff
@@ -168,92 +172,151 @@ export class DemexWallet {
+    this.grantee = grantee;
+  }
 
-    try {
-      // retrieve account info, reload if sequenceInvalidated flag is set.
-      if (!this.accountInfo
-        || this.accountInfo?.address === this.evmBech32Address // refresh to check if carbon acc is present
-        || this.sequenceInvalidated)
-        await this.reloadAccountSequence();
-
-      // prepare tx timeout height. add a default timeout height to tx if unset.
-      // skip if EIP-712 signer, timeoutHeight is not supported for EIP-712
-      let timeoutHeight;
-      if (!isDemexEIP712Signer(this.signer)) {
-        timeoutHeight = await this.getTimeoutHeight();
-      }
 
-      // calculate tx sequence number, will be 0 if accountInfo doesnt exist (new account).
-      // then increment account info sequence number if not a new account.
-      let sequence: number
-      if (!this.accountInfo) {
-        sequence = signOpts?.sequence ?? 0
-      } else {
-        sequence = this.accountInfo!.sequence;
-        this.accountInfo = {
-          ...this.accountInfo!,
-          sequence: sequence + 1,
-        };
-      }
+  private useGrantee(messages: readonly EncodeObject[], bypass?: boolean) {
+    const typeUrls = new Set(messages.map(m => m.typeUrl))
+    return !!this.grantee && this.grantee.isAuthorised(typeUrls) && !bypass
+  }
 
-      // obtain signed transaction
-      const _signOpts: SignTxOpts = {
-        ...signOpts,
-        explicitSignerData: {
-          timeoutHeight,
-          ...signOpts?.explicitSignerData,
-        },
-      };
-      const signedTx = await this.getSignedTx(signerAddress, messages, sequence, _signOpts);
+  private async getGranteeSigningData(txRequest: SignTxRequest) {
+    const { messages } = txRequest;
+    return {
+      ...txRequest,
+      signOpts: { ...txRequest.signOpts, feeGranter: this.bech32Address },
+      address: await this.grantee!.getAddress(),
+      messages: [await this.grantee!.constructExecMessage(messages)],
+      signer: this.grantee!.signer,
+      signingClient: await this.grantee!.getSigningClient(await this.getTmClient()),
+    }
+  }
+
+  private async getDefaultSigningData(txRequest: SignTxRequest) {
+    return {
+      ...txRequest,
+      address: this.bech32Address,
+      evmBech32Address: this.evmBech32Address,
+      signer: this.signer,
+      signingClient: await this.getSigningStargateClient(),
+    }
+  }
+
+  private async getSigningData(txRequest: SignTxRequest): Promise<SigningData> {
+    const { messages, signOpts } = txRequest;
+    if (this.useGrantee(messages, signOpts?.bypassGrantee)) return await this.getGranteeSigningData(txRequest);
+    return await this.getDefaultSigningData(txRequest);
+  }
+
+  private async checkReloadAccountState(reloadAddresses: ReloadAddresses) {
+    const { address } = reloadAddresses
+    const state = this.accountStates?.[address]
+    if (!state || state.sequenceInvalidated) return await this.reloadAccount(reloadAddresses)
+  }
```

Comment:
recommend passing in the key used to index `this.accountStates` instead of custom `ReloadAddresses` type.

Reply:
i also passed in `signer` cos it may be an account init with an evmaddress first. have to use signer to get the evmaddress. didnt wanna use this.evmbech32 to stay true to the source (ie the signer like metamask/rainbowkit)